airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/global_substream_cursor.py

```diff
@@ -0,0 +1,277 @@
+
+
+class Timer:
+    """
+    A simple timer class that measures elapsed time in seconds using a high-resolution performance counter.
+    """
+
+    def __init__(self) -> None:
+        self._start: Optional[int] = None
+
+    def start(self) -> None:
+        self._start = time.perf_counter_ns()
+
+    def finish(self) -> int:
+        if self._start:
+            return int((time.perf_counter_ns() - self._start) // 1e9)
+        else:
+            raise RuntimeError("Global substream cursor timer not started")
+
+    @property
+    def is_started(self) -> bool:
+        return self._start is not None
+
+
+class GlobalCursorStreamSlice(StreamSlice):
+    """
+    A stream slice that is used to track the state of substreams using a single global cursor.
+    This class has the additional `last_slice` field that is used to indicate if the stream slice is the last.
+    """
+
+    def __init__(self, *, partition: Mapping[str, Any], cursor_slice: Mapping[str, Any], last_slice: bool = False) -> None:
+        super().__init__(partition=partition, cursor_slice=cursor_slice)
+        self._last_slice = last_slice
+
+    @property
+    def last_slice(self) -> bool:
+        return self._last_slice
+
+
+class GlobalSubstreamCursor(DeclarativeCursor):
+    """
+    The GlobalSubstreamCursor is designed to track the state of substreams using a single global cursor.
+    This class is beneficial for streams with many partitions, as it allows the state to be managed globally
+    instead of per partition, simplifying state management and reducing the size of state messages.
+
+    This cursor is activated by setting the `global_substream_cursor` parameter for incremental sync.
+
+    Warnings:
+    - This class enforces a minimal lookback window for substream based on the duration of the previous sync to avoid losing records. This lookback ensures that any records added or updated during the sync are captured in subsequent syncs.
+    - The global cursor is updated only at the end of the sync. If the sync ends prematurely (e.g., due to an exception), the state will not be updated.
+    - When using the `incremental_dependency` option, the sync will progress through parent records, preventing the sync from getting infinitely stuck. However, it is crucial to understand the requirements for both the `global_substream_cursor` and `incremental_dependency` options to avoid data loss.
+    """
+
+    def __init__(self, stream_cursor: DeclarativeCursor, partition_router: PartitionRouter):
+        self._stream_cursor = stream_cursor
+        self._partition_router = partition_router
+        self._timer = Timer()
+        self._lookback_window: Optional[int] = None
+
+    def stream_slices(self) -> Iterable[GlobalCursorStreamSlice]:
+        """
+        Generate stream slices and flag the last slice.
+        """
+
+        def flag_last(generator: Generator[StreamSlice, None, None]) -> Generator[GlobalCursorStreamSlice, None, None]:
```

Comment:
nit: mark_as_last_slice

Reply:
This function yields every slice, but only the last slice is flagged

Reply:
Yes, as far as it technically `gets` the last slice from whatever passed into it - it's not only flags but also returns the `last_slice`, so for me it's easier to follow the flow with the name like that.